pkg/auth/providers/saml/saml_actions.go

```diff
@@ -0,0 +1,96 @@
+package saml
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/rancher/norman/httperror"
+	"github.com/rancher/norman/types"
+	"github.com/rancher/rancher/pkg/auth/providers/common"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+)
+
+func (s *Provider) formatter(apiContext *types.APIContext, resource *types.RawResource) {
+	common.AddCommonActions(apiContext, resource)
+	resource.AddAction(apiContext, "configureTest")
+	resource.AddAction(apiContext, "testAndApply")
+}
+
+func (s *Provider) actionHandler(actionName string, action *types.Action, request *types.APIContext) error {
+	handled, err := common.HandleCommonAction(actionName, action, request, s.Name, s.AuthConfigs)
+	if err != nil {
+		return err
+	}
+	if handled {
+		return nil
+	}
+
+	if actionName == "configureTest" {
+		return s.configureTest(actionName, action, request)
+	} else if actionName == "testAndApply" {
+		return s.testAndApply(actionName, action, request)
+	}
+
+	return httperror.NewAPIError(httperror.ActionNotAvailable, "")
+}
+
+func (s *Provider) configureTest(actionName string, action *types.Action, request *types.APIContext) error {
+	samlConfig := &v3.SamlConfig{}
+	if err := json.NewDecoder(request.Request.Body).Decode(samlConfig); err != nil {
+		return httperror.NewAPIError(httperror.InvalidBodyContent,
+			fmt.Sprintf("SAML: Failed to parse body: %v", err))
+	}
+
+	redirectURL := s.formSamlRedirectURL(samlConfig)
+	err := InitializeSamlServiceProvider(samlConfig, s.Name)
+	if err != nil {
+		return fmt.Errorf("SAML: error in initializing saml client: %v", err)
+	}
+
+	data := map[string]interface{}{
+		"redirectUrl": redirectURL,
+		"type":        "samlConfigTestOutput",
+	}
+
+	annotations := make(map[string]string)
+	annotations["configured"] = "true"
+	annotations["userID"] = s.UserMGR.GetUser(request)
```

Comment:
saving the userID as an annotation is bad idea. we need to find a better way

Reply:
set the userID as a claim on the signed jwt cookie that we are using for "state"